### ISSUE TITLE
dql_grasping: Training with default config did not converge to good success rate

### ISSUE THREAD

**Issue by kingwayliang** [google-research/google-research#14]
We tried to recreate some of the results in dql_grasping. After setting up the environment according to requirements, we ran run_random_collect_oss.sh, and then run_train_collect_eval_oss.sh with dqn on-policy and dqn off-policy. The results shown in images below suggest the training didn't converge on policies with expected success rate, what steps should we take to reproduce similar results to those presented in the paper?

![eval1](https://user-images.githubusercontent.com/46766340/52463626-74d23780-2b2c-11e9-9be0-aa4dca945e5d.png)
![eval2](https://user-images.githubusercontent.com/46766340/52463630-78fe5500-2b2c-11e9-8d24-b714ba4aeff6.png)
![test1](https://user-images.githubusercontent.com/46766340/52463633-7b60af00-2b2c-11e9-875f-db651344326e.png)
![test2](https://user-images.githubusercontent.com/46766340/52463635-7e5b9f80-2b2c-11e9-81c9-511bc1e53c73.png)

**Comment by ericjang** [comment]
Good performance results (for both on/off-policy) require seeding the training with quite a bit of data collected with the random policy before training. Try collecting at least 100k episodes of random policy first (it is really important to run this collection in parallel on a CPU cluster, otherwise gathering this data takes a long time).

**Comment by andreykurenkov** [comment]
@ericjang continuing discussion here as per email. 

After collecting 100k grasps as you suggested , we got ~0.55 final test perf for both on and off policy. We are hoping to match the results in your paper of 0.7 & 0.8 performance as in your paper.

We don't have infra to distribute over many CPUs, so would rather not try with 1M - since your provide results for 100k it seems fine to try and replicate those. We don't modify any other aspects of the configs in the repo. Could you suggest any other things to try or check to get replication working?

**Comment by ericjang** [comment]
We re-ran our Q-learning  experiments with replicated trials on P100 gpus - here's a screenshot of the tensorboard.
![image](https://user-images.githubusercontent.com/433201/53602136-ba4db780-3b62-11e9-956e-ef56cba47215.png)

To be honest I haven't run this code in a non-Google environment (e.g. on Google cloud), perhaps a difference could arise from there. One other thing to try - when gathering 100k episodes, we distribute across 1000 different collect workers, each which resets with a different random seed.

**Comment by andreykurenkov** [comment]
Are all the different lines different runs? The variance looks to be pretty high (at a glance, higher than as shown in paper, though ofc hard to tell), your lowest outcome (orange) is close to our outcome.

We will try scaling data collection to just 10 CPUs + running a few times, see what we get.

**Comment by kingwayliang** [comment]
We re-collected 100k episodes using 8 CPUs and re-trained with on-policy, got the results below.

![on_eval2](https://user-images.githubusercontent.com/46766340/53850178-32efb200-3f6f-11e9-8a88-78ea94db663f.png)
![on_test2](https://user-images.githubusercontent.com/46766340/53850182-35eaa280-3f6f-11e9-939f-72543daaa76b.png)

Both eval and test rewards peaked at 250~300k training steps but subsequently decreased to values similar to our previous results. @ericjang the re-ran results you posted show training with 200k steps, while we've been training with the default 2M steps, can this be the problem? It seems strange the policy would get worse after more training steps.

We also re-trained off-policy using the new episodes and only for 200k steps, with the following results. Test reward is still relatively low.

![off_eval2](https://user-images.githubusercontent.com/46766340/53850687-0046b900-3f71-11e9-9e87-08eb3abb4aa3.png)
![off_test2](https://user-images.githubusercontent.com/46766340/53850690-050b6d00-3f71-11e9-9f81-97a586c3cdc9.png)

**Comment by kingwayliang** [comment]
After more experimentation, we found:

1. having a max number of training steps 2x to 2.5x of the number of collected episodes gives the best training result.

2. The grasping environment selects random objects to spawn the first time the environment runs setup and doesn't re-select after. So if episodes are collected on small number of CPU's, the environment should be tweaked to re-select objects by calling _get_random_objects to ensure episodes cover a diverse set of objects.

3. The urdf_pattern defined in _get_random_objects for training object did not work properly on our machine and only picked out 100 objects. Changing [^0] to [!0] solved the problem for us.

With the above fixes we were able to reproduce results similar to those in the paper.
